src/pages/Admin/MapPreview/MapPreview.tsx

```diff
@@ -58,25 +72,27 @@ const MapPreview: React.FC<MapPreviewProps> = ({
     }));
   }, [isEditing, mapRooms, rooms]);
 
-  if (!user) return <></>;
-
-  const handleOnChange = (val: SubVenueIconMap) => {
-    if (!isEqual(roomRef.current, val)) {
-      roomRef.current = val;
-      const normalizeRooms = Object.values(val).map((room, index) => ({
-        ...rooms[index],
-        x_percent: room.left,
-        y_percent: room.top,
-        width_percent: room.width,
-        height_percent: room.height,
-      }));
-      setMapRooms(normalizeRooms);
-      onRoomChange && onRoomChange(normalizeRooms);
-    }
-  };
+  const updateRoomPosition = useCallback(
+    (val: SubVenueIconMap) => {
+      if (!isEqual(roomRef.current, val)) {
+        roomRef.current = val;
+        const normalizeRooms = Object.values(val).map((room, index) => ({
+          ...rooms[index],
+          x_percent: room.left,
+          y_percent: room.top,
+          width_percent: room.width,
+          height_percent: room.height,
+        }));
+        setMapRooms(normalizeRooms);
+        onRoomChange && onRoomChange(normalizeRooms);
+      }
+    },
+    [onRoomChange, rooms]
+  );
+
+  const [{ loading: isSaving }, saveRoomPositions] = useAsyncFn(async () => {
+    if (isSaving || !user) return;
 
-  const handleSavePositions = async () => {
```

Comment:
Due to UI limitations, suggestion:

```
    // more descriptive name, could be boxModels or any similar combination
    const iconsMapBoxes = Object.values(roomRef.current);

    let roomIndex = 0;

    // no need of extra token variable
    for (const { height, left, top, width } of iconsMapBoxes) {
      const roomInput: RoomInput_v2 = {
        ...rooms[roomIndex],
        x_percent: left,
        y_percent: top,
        width_percent: width,
        height_percent: height,
      };

      await updateRoom(roomInput, venueId, user, roomIndex);

      // ++ has no performance benefit over +=1 in JS while +=1 as a style shows unambiguous intent
      roomIndex += 1;
    }
```

Reply:
Fixed in: https://github.com/sparkletown/sparkle/pull/2006/commits/0c72b4d0e7a6245b75d442138529cfd256342c4b